Rubberduck.Parsing/VBA/ReferenceResolveRunnerBase.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Rubberduck.VBEditor;
+using Rubberduck.Parsing.Symbols;
+using Antlr4.Runtime.Tree;
+using System.Diagnostics;
+using NLog;
+
+namespace Rubberduck.Parsing.VBA
+{
+    public abstract class ReferenceResolveRunnerBase : IReferenceResolveRunner
+    {
+
+        //This is used to buffer the modules for which to resolve references
+        //to avoid missing modules in case of a cancellation or a resolver error.
+        private readonly HashSet<QualifiedModuleName> _toResolve = new HashSet<QualifiedModuleName>();
+
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+        protected readonly RubberduckParserState _state;
+        protected readonly IParserStateManager _parserStateManager;
+        private readonly IModuleToModuleReferenceManager _moduleToModuleReferenceManager;
+
+
+        public ReferenceResolveRunnerBase(
+            RubberduckParserState state,
+            IParserStateManager parserStateManager,
+            IModuleToModuleReferenceManager moduletToModuleReferenceManager)
+        {
+            if (state == null) throw new ArgumentNullException(nameof(state));
+            if (parserStateManager == null) throw new ArgumentNullException(nameof(parserStateManager));
+            if (moduletToModuleReferenceManager == null) throw new ArgumentNullException(nameof(moduletToModuleReferenceManager));
+
+            _state = state;
+            _parserStateManager = parserStateManager;
+            _moduleToModuleReferenceManager = moduletToModuleReferenceManager;
+        }
+
+
+        protected abstract void ResolveReferences(ICollection<KeyValuePair<QualifiedModuleName, IParseTree>> toResolve, CancellationToken token);
+        protected abstract void AddModuleToModuleReferences(DeclarationFinder finder, CancellationToken token);
+
+
+        public void ResolveReferences(ICollection<QualifiedModuleName> toResolve, CancellationToken token)
+        {
+                token.ThrowIfCancellationRequested();
```

Comment:
The reason why `token.ThrowIfCancellationRequested();` is indented one further level is that it really hurts readibility of the actual program logic when it is on the same level. However, if you insist, I will put it on the same level.